Wikipedia/Code/WMFColumnarCollectionViewLayout.m

```diff
@@ -0,0 +1,165 @@
+#import "WMFColumnarCollectionViewLayout.h"
+#import "WMFCVLInfo.h"
+#import "WMFCVLColumn.h"
+#import "WMFCVLSection.h"
+#import "WMFCVLAttributes.h"
+#import "WMFCVLInvalidationContext.h"
+#import "WMFCVLMetrics.h"
+
+@interface WMFColumnarCollectionViewLayout ()
+
+@property (nonatomic, readonly) id <WMFColumnarCollectionViewLayoutDelegate> delegate;
+@property (nonatomic, strong) WMFCVLMetrics *metrics;
+@property (nonatomic, strong) WMFCVLInfo *info;
+@property (nonatomic) BOOL needsNewLayout;
+@property (nonatomic) BOOL needsLayout;
+
+@end
+
+@implementation WMFColumnarCollectionViewLayout
+
+- (nonnull instancetype)initWithMetrics:(nonnull WMFCVLMetrics *)metrics {
+    self = [super init];
+    if (self) {
+        self.metrics = metrics;
+        self.needsNewLayout = YES;
+        self.needsLayout = NO;
+    }
+    return self;
+}
+
+- (instancetype)init {
+    return [self initWithMetrics:[WMFCVLMetrics defaultMetrics]];
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self = [super initWithCoder:aDecoder];
+    if (self) {
+        if (!self.metrics) {
+            self.metrics = [WMFCVLMetrics defaultMetrics];
+        }
+    }
+    return self;
+}
+
+#pragma mark - Classes
+
++ (Class)invalidationContextClass {
+    return [WMFCVLInvalidationContext class];
+}
+
++ (Class)layoutAttributesClass {
+    return [WMFCVLAttributes class];
+}
+
+#pragma mark - Properties
+
+- (id <WMFColumnarCollectionViewLayoutDelegate>)delegate {
+    assert(self.collectionView.delegate == nil || [self.collectionView.delegate conformsToProtocol:@protocol(WMFColumnarCollectionViewLayoutDelegate)]);
+    return (id <WMFColumnarCollectionViewLayoutDelegate>)self.collectionView.delegate;
+}
+
+- (NSInteger)numberOfItemsInSection:(NSInteger)section {
+    return [self.collectionView.dataSource collectionView:self.collectionView numberOfItemsInSection:section];
+}
+
+- (CGSize)collectionViewContentSize {
+    return self.info.contentSize;
+}
+
+#pragma mark - Layout
+
+- (nullable NSArray<__kindof UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect {
+    
+    NSMutableArray *attributesArray = [NSMutableArray array];
+    
+    [self.info enumerateSectionsWithBlock:^(WMFCVLSection * _Nonnull section, NSUInteger idx, BOOL * _Nonnull stop) {
```

Comment:
Should this be encapsulated in the Info class so you can implement this method in one line similarly to the calls below?